### PR TITLE
Implement scrollbar-gutter both-edges support on non-viewports

### PR DESCRIPTION
**Pull request by lukewarlow**
#### 9a055c133ed578da42d94db83e7e62e367f0e25f
<pre>
Implement scrollbar-gutter both-edges support on non-viewports
<a href="https://bugs.webkit.org/show_bug.cgi?id=257708">https://bugs.webkit.org/show_bug.cgi?id=257708</a>

Reviewed by Simon Fraser.

Scrollbar gutter now correctly works (on non-viewports) for both the
&quot;stable&quot; and &quot;stable both-edges&quot; values.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/scrollbar-gutter-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/scrollbar-gutter-rtl-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/scrollbar-gutter-vertical-lr-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/scrollbar-gutter-vertical-rl-001-expected.txt:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::determineLogicalLeftPositionForChild):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::contentBoxLocation const):
* Source/WebCore/rendering/RenderBoxInlines.h:
(WebCore::RenderBox::computedCSSContentBoxRect const):
(WebCore::RenderBox::contentHeight const):
(WebCore::RenderBox::contentWidth const):
(WebCore::RenderBox::overridingContentLogicalHeight const):
(WebCore::RenderBox::overridingContentLogicalWidth const):
* Source/WebCore/rendering/style/RenderStyle.h:

Canonical link: <a href="https://commits.webkit.org/265199@main">https://commits.webkit.org/265199@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/7efe6a9cf19ba2dafaeebfa07197d264e85db2bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10197 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10434 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11842 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9838 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10208 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12428 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10386 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12770 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10353 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11100 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8596 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12225 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8405 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9235 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16528 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8609 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9504 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9387 "1 flakes 3 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12634 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9673 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9844 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7992 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10330 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9003 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9130 "Exiting early after 10 failures. 6071 tests run. 1 flakes") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2671 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2446 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13251 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10611 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9671 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2628 "Passed tests") |

### CODE REVIEW

**Comment by lukewarlow** [comment]
Blocked by https://github.com/WebKit/WebKit/pull/14598

**Comment by lukewarlow** [comment]
Code still needs cleaning up.

` - scrollbarLogicalHeight() - (style().scrollbarGutter().bothEdges ? scrollbarLogicalHeight() : 0)`

Might be better as the below for example

` - (scrollbarLogicalHeight() * (style().scrollbarGutter().bothEdges ? 2 : 1))`

File: Source/WebCore/rendering/RenderBox.cpp [hunk 1]
```diff
@@ -762,8 +762,9 @@ LayoutRect RenderBox::paddingBoxRect() const
 
 LayoutPoint RenderBox::contentBoxLocation() const
 {
-    LayoutUnit scrollbarSpace = shouldPlaceVerticalScrollbarOnLeft() ? verticalScrollbarWidth() : 0;
-    return { borderLeft() + paddingLeft() + scrollbarSpace, borderTop() + paddingTop() };
+    LayoutUnit verticalScrollbarSpace = (shouldPlaceVerticalScrollbarOnLeft() || style().scrollbarGutter().bothEdges) ? verticalScrollbarWidth() : 0;
```
Comment (smfr) [inline Source/WebCore/rendering/RenderBox.cpp:765]: These additions of `scrollbarGutter()` everywhere are making the code pretty hard to follow. I think it would be much nicer to add something like `verticalScrollbarSpace()` which consults `verticalScrollbarWidth()` and `scrollbarGutter()`.

Comment (lukewarlow) [reply Source/WebCore/rendering/RenderBox.cpp:765]: @smfr 

So I've had a look at this and while I agree the code could probably be cleaned up I don't fully understand what you mean by this. the "scrollbarSpace" is only used in the 1 place.

To me its RenderBoxInlines that probably needs cleaning up the most but again I'm unsure how best to do so.

Comment (lukewarlow) [reply Source/WebCore/rendering/RenderBox.cpp:765]: This extra scrollbar gutter check along with the ones in RenderBoxInlines are specifically for handling the "scrollbar-gutter: stable both-edges" situation.

As shown in the below image there's an overlap between RTL scrollbars and scrollbar-gutter both-edges which is why this check was added here.

image

**Review by smfr — APPROVED** [review]
